### PR TITLE
Fixes log deduplication when mutating Labels using LogQL

### LINKED ISSUE #4500: Deduping with label_format or metrics aggregation can be wrong
When using labels aggregation or labels format the original labelset from the chunk is changed.

However currently we use the labels to identify different chunks when deduping entries/sample that have the same timestamp.

see 

https://github.com/grafana/loki/blob/6ac96478a7704a0d004888b9ca3e878c7d909d47/pkg/iter/sample_iterator.go#L219

and 

https://github.com/grafana/loki/blob/6ac96478a7704a0d004888b9ca3e878c7d909d47/pkg/iter/entry_iterator.go#L230

I suggest we introduce a new functions to both iterator interfaces `LabelHash()` for now to not affect performance we could use the fingerprint for the chunk.

We should then rely on this new function for deduplication.

### PR DESCRIPTION
**Pull request by cyriltovena**
**What this PR does / why we need it**:

Labels for each sample and log were used to find and remove duplicates when iterating over them. However since those labels can be mutated by the LogQL engine for optimization or by the users ( `label_format`, `sum()`) the deduplication wasn't consistent anymore.

This introduce a new StreamHash function that contains the original stream label hash and can be used for deduplication purpose. This ensure the deduplication is consistent and independent of the queries made.

This PR is built on https://github.com/grafana/loki/pull/5281 which separate the sorting logic to the dedupe logic allowing to have different behaviour, this was required to make sharding work.

**Which issue(s) this PR fixes**:
Fixes #4500

Example of metric queries with duplicates before and after:

image

The last hour is still buggy because I did not deploy the ingesters during my test.

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/chunkenc/memchunk.go [hunk 1]
```diff
@@ -1299,6 +1301,8 @@ func (e *sampleBufferedIterator) Next() bool {
 }
 func (e *sampleBufferedIterator) Labels() string { return e.currLabels.String() }
 
+func (e *sampleBufferedIterator) StreamHash() uint64 { return e.extractor.BaseLabels().Hash() }
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/chunkenc/memchunk.go:1304]: The fingerprint used is legacy from cortex, but is ultimatily also a hash value of the labels. However the legacy fingerprint is using fnv while the new hash is using xxhash2, the latter is not only faster but also way less chance of collision.

I named it StreamHash to be independent of the implementation, here we're using xxhash2.

**Review by owen-d — APPROVED** [review]

File: pkg/iter/sample_iterator_test.go [hunk 2]
```diff
@@ -170,13 +220,14 @@ func TestNewSampleQueryClientIterator(t *testing.T) {
 }
 
 func TestNewNonOverlappingSampleIterator(t *testing.T) {
+	// todo fix this test
```
Comment (sandeepsukhani) [inline pkg/iter/sample_iterator_test.go:223]: Can you please elaborate more in this comment for future reference?

**Review by sandeepsukhani — APPROVED** [review]
LGTM! Nice work!

File: pkg/iter/sample_iterator_test.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/iter/sample_iterator_test.go:223]: I guess I forgot about that I did fix it. :)

I add a lot of comment during the large refactoring to remember what to fix and that's a left over.

File: pkg/iter/entry_iterator.go [hunk 3]
```diff
@@ -86,44 +72,24 @@ func (h *iteratorHeap) Pop() interface{} {
...
-	switch {
-	case un1 < un2:
-		return true
-	case un1 > un2:
-		return false
-	default: // un1 == un2:
-		return h.iteratorHeap[i].Labels() < h.iteratorHeap[j].Labels()
+func (h iteratorSortHeap) Less(i, j int) bool {
+	t1, t2 := h.iteratorHeap[i].Entry().Timestamp.UnixNano(), h.iteratorHeap[j].Entry().Timestamp.UnixNano()
+	if t1 == t2 {
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/iter/entry_iterator.go:83]: I was about to do so and realized performance wise doing nano is a bit better

**Pull request merged**